File: plugins/plugin-xai/typescript/post.ts
import {
  ChannelType,
  createUniqueUuid,
  type IAgentRuntime,
  logger,
  type Memory,
  ModelType,
  parseBooleanFromText,
  type UUID,
} from "@elizaos/core";
import type { ClientBase } from "./base";
import { getRandomInterval } from "./environment";
import type { MediaData, PostResponse } from "./types";
import { sendPost } from "./utils";
import {
  addToRecentPosts,
  createMemorySafe,
  ensureXContext,
  isDuplicatePost,
} from "./utils/memory";
import { getSetting } from "./utils/settings";
/**
 * Class representing an X post client for generating and posting.
 */
export class XPostClient {
  client: ClientBase;
  runtime: IAgentRuntime;
  xUsername: string;
  private isDryRun: boolean;
  private state: Record<string, unknown>;
  private isRunning: boolean = false;
  private isPosting: boolean = false; // Add lock to prevent concurrent posting

  /**
   * Creates an instance of XPostClient.
   * @param {ClientBase} client - The client instance.
   * @param {IAgentRuntime} runtime - The runtime instance.
   * @param {Record<string, unknown>} state - The state object containing configuration settings
   */
  constructor(client: ClientBase, runtime: IAgentRuntime, state: Record<string, unknown>) {
    this.client = client;
    this.state = state;
    this.runtime = runtime;
    const dryRunSetting =
      typeof this.state?.X_DRY_RUN === "string" ||
      typeof this.state?.X_DRY_RUN === "boolean" ||
      this.state?.X_DRY_RUN === true ||
      this.state?.X_DRY_RUN === false
        ? this.state.X_DRY_RUN
        : getSetting(this.runtime, "X_DRY_RUN");
    this.isDryRun = parseBooleanFromText(dryRunSetting);

    // Get X username from settings
    const usernameSetting = getSetting(this.runtime, "X_USERNAME") || this.state?.X_USERNAME;
    this.xUsername = typeof usernameSetting === "string" ? usernameSetting : "";

    // Log configuration on initialization
    logger.log("X Post Client Configuration:");
    logger.log(`- Dry Run Mode: ${this.isDryRun ? "Enabled" : "Disabled"}`);

    const postIntervalMin = parseInt(
      (typeof this.state?.X_POST_INTERVAL_MIN === "string"
        ? this.state.X_POST_INTERVAL_MIN
        : null) ||
        (getSetting(this.runtime, "X_POST_INTERVAL_MIN") as string) ||
        "90",
      10
    );
    const postIntervalMax = parseInt(
      (typeof this.state?.X_POST_INTERVAL_MAX === "string"
        ? this.state.X_POST_INTERVAL_MAX
        : null) ||
        (getSetting(this.runtime, "X_POST_INTERVAL_MAX") as string) ||
        "150",
      10
    );
    logger.log(`- Post Interval: ${postIntervalMin}-${postIntervalMax} minutes (randomized)`);
  }

  /**
   * Stops the X post client
   */
  async stop() {
    logger.log("Stopping X post client...");
    this.isRunning = false;
  }

  /**
   * Starts the X post client, setting up a loop to periodically generate new posts.
   */
  async start() {
    logger.log("Starting X post client...");
    this.isRunning = true;

    const generateNewPostLoop = async () => {
      if (!this.isRunning) {
        logger.log("X post client stopped, exiting loop");
        return;
      }

      await this.generateNewPost();

      if (!this.isRunning) {
        logger.log("X post client stopped after post, exiting loop");
        return;
      }

      // Get random post interval in minutes
      const postIntervalMinutes = getRandomInterval(this.runtime, "post");

      // Convert to milliseconds
      const interval = postIntervalMinutes * 60 * 1000;

      logger.info(`Next post scheduled in ${postIntervalMinutes.toFixed(1)} minutes`);

      // Wait for the interval AFTER generating the post
      await new Promise((resolve) => setTimeout(resolve, interval));

      if (this.isRunning) {
        // Schedule the next iteration
        generateNewPostLoop();
      }
    };

    // Wait a bit longer to ensure profile is loaded
    await new Promise((resolve) => setTimeout(resolve, 5000));

    // Check if we should generate a post immediately
    const postImmediately =
      typeof this.state?.X_POST_IMMEDIATELY === "string" ||
      typeof this.state?.X_POST_IMMEDIATELY === "boolean"
        ? this.state.X_POST_IMMEDIATELY
        : (getSetting(this.runtime, "X_POST_IMMEDIATELY") as string);

    if (parseBooleanFromText(postImmediately)) {
      logger.info("X_POST_IMMEDIATELY is true, generating initial post now");
      // Try multiple times in case profile isn't ready
      let retries = 0;
      while (retries < 5) {
        const success = await this.generateNewPost();
        if (success) break;

        retries++;
        logger.info(`Retrying immediate post (attempt ${retries}/5)...`);
        await new Promise((resolve) => setTimeout(resolve, 3000));
      }
    }

    // Start the regular generation loop
    generateNewPostLoop();
  }

  /**
   * Handles the creation and posting of a post by emitting standardized events.
   * This approach aligns with our platform-independent architecture.
   * @returns {Promise<boolean>} true if post was posted successfully
   */
  async generateNewPost(): Promise<boolean> {
    logger.info("Attempting to generate new post...");

    // Prevent concurrent posting
    if (this.isPosting) {
      logger.info("Already posting, skipping concurrent attempt");
      return false;
    }

    this.isPosting = true;

    try {
      // Create the timeline room ID for storing the post
      const userId = this.client.profile?.id;
      if (!userId) {
        logger.error("Cannot generate post: X profile not available");
        this.isPosting = false; // Reset flag
        return false;
      }

      logger.info(`Generating post for user: ${this.client.profile?.username} (${userId})`);

      // Create standardized world and room IDs
      const _worldId = createUniqueUuid(this.runtime, userId) as UUID;
      const roomId = createUniqueUuid(this.runtime, `${userId}-home`) as UUID;

      // Generate post content using the runtime's model
      const state = await this.runtime
        .composeState({
          agentId: this.runtime.agentId,
          entityId: this.runtime.agentId,
          roomId,
          content: { text: "", type: "post" },
          createdAt: Date.now(),
        } as Memory)
        .catch((error) => {
          logger.warn("Error composing state, using minimal state:", error);
          // Return minimal state if composition fails
          return {
            agentId: this.runtime.agentId,
            recentMemories: [],
            values: {},
          };
        });

      // Create a prompt for post generation
      const postPrompt = `You are ${this.runtime.character.name}.
${this.runtime.character.bio}

CRITICAL: Generate a post that sounds like YOU, not a generic motivational poster or LinkedIn influencer.

${
  this.runtime.character.messageExamples && this.runtime.character.messageExamples.length > 0
    ? `
Example posts that capture your voice:
${(
  this.runtime.character.messageExamples as Array<{
    examples?: Array<{ content?: { text?: string } }>;
  }>
)
  .flatMap((group) => group.examples ?? [])
  .map((example) => example.content?.text ?? "")
  .filter((text) => text.length > 0)
  .slice(0, 5)
  .join("\n")}
`
    : ""
}

Style guidelines:
- Be authentic, opinionated, and specific - no generic platitudes
- Use your unique voice and perspective
- Share hot takes, unpopular opinions, or specific insights
- Be conversational, not preachy
- If you use emojis, use them sparingly and purposefully
- Length: 50-280 characters (keep it punchy)
- NO hashtags unless absolutely essential
- NO generic motivational content

Your interests: ${this.runtime.character.topics?.join(", ") || "technology, crypto, AI"}

${
  this.runtime.character.style
    ? `Your style: ${
        typeof this.runtime.character.style === "object"
          ? this.runtime.character.style.all?.join(", ") ||
            JSON.stringify(this.runtime.character.style)
          : this.runtime.character.style
      }`
    : ""
}

Recent context:
${
  Array.isArray(state.recentMemories) && state.recentMemories.length > 0
    ? state.recentMemories
        .slice(0, 3)
        .map((m: Memory) => m.content?.text || "")
        .join("\n") || "No recent context"
    : "No recent context"
}

Generate a single post that sounds like YOU would actually write it:`;

      // Use the runtime's model to generate post content
      const generatedContent = await this.runtime.useModel(ModelType.TEXT_SMALL, {
        prompt: postPrompt,
        temperature: 0.9, // Increased for more creativity
        maxTokens: 100,
      });

      const postText = generatedContent.trim();

      if (!postText || postText.length === 0) {
        logger.error("Generated empty post content");
        return false;
      }

      if (postText.includes("Error: Missing")) {
        logger.error("Error in generated content:", postText);
        return false;
      }

      // Validate post length
      if (postText.length > 280) {
        logger.warn(`Generated post too long (${postText.length} chars), truncating...`);
        // Truncate to the last complete sentence within 280 chars
        const sentences = postText.match(/[^.!?]+[.!?]+/g) || [postText];
        let truncated = "";
        for (const sentence of sentences) {
          if ((truncated + sentence).length <= 280) {
            truncated += sentence;
          } else {
            break;
          }
        }
        const finalPost = truncated.trim() || `${postText.substring(0, 277)}...`;
        logger.info(`Truncated post: ${finalPost}`);

        // Post the truncated post
        if (this.isDryRun) {
          logger.info(`[DRY RUN] Would post: ${finalPost}`);
          return false;
        }

        const result = await this.postToX(finalPost, []);

        if (result === null) {
          logger.info("Skipped posting duplicate post");
          return false;
        }

        const postId = result.id ?? result.data?.id ?? result.data?.data?.id;
        logger.info(`Post created successfully! ID: ${postId}`);

        // Don't save to memory if room creation might fail
        logger.info("Post created successfully (memory saving disabled due to room constraints)");
        return true;
      }

      logger.info(`Generated post: ${postText}`);

      // Post the post
      if (this.isDryRun) {
        logger.info(`[DRY RUN] Would post: ${postText}`);
        return false;
      }

      const result = await this.postToX(postText, []);

      // If result is null, it means we detected a duplicate post and skipped posting
      if (result === null) {
        logger.info("Skipped posting duplicate post");
        return false;
      }

      const postId = result.id ?? result.data?.id ?? result.data?.data?.id;
      logger.info(`Post created successfully! ID: ${postId}`);

      if (result && postId) {
        const postedPostId = createUniqueUuid(this.runtime, postId);

        try {
          // Ensure context exists with error handling
          const context = await ensureXContext(this.runtime, {
            userId,
            username: this.client.profile?.username || "unknown",
            conversationId: `${userId}-home`,
          });

          // Create memory for the posted post with retry logic
          const postedMemory: Memory = {
            id: postedPostId,
            entityId: this.runtime.agentId,
            agentId: this.runtime.agentId,
            roomId: context.roomId,
            content: {
              text: postText,
              source: "x",
              channelType: ChannelType.FEED,
              type: "post",
              metadata: {
                postId,
                postedAt: Date.now(),
              },
            },
            createdAt: Date.now(),
          };

          await createMemorySafe(this.runtime, postedMemory, "messages");
          logger.info("Post created and saved to memory successfully");
        } catch (error) {
          logger.error(
            "Failed to save post memory:",
            error instanceof Error ? error.message : String(error)
          );
          // Don't fail the post creation if memory creation fails
        }

        return true;
      } else {
        logger.warn("Post generation returned no result");
        return false;
      }
    } catch (error) {
      logger.error(
        "Error generating post:",
        error instanceof Error ? error.message : String(error)
      );
      return false;
    } finally {
      this.isPosting = false;
    }
  }

  /**
   * Posts content to X
   * @param {string} text The post text to create
   * @param {MediaData[]} mediaData Optional media to attach to the post
   * @returns {Promise<PostResponse | null>} The result from the X API
   */
  private async postToX(text: string, mediaData: MediaData[] = []): Promise<PostResponse | null> {
    // Check if this post is a duplicate of recent posts
    const username = this.client.profile?.username;
    if (!username) {
      logger.error("No profile username available");
      return null;
    }

    // Check for duplicates in recent posts
    const isDuplicate = await isDuplicatePost(this.runtime, username, text);
    if (isDuplicate) {
      logger.warn("Post is a duplicate of a recent post. Skipping to avoid duplicate.");
      return null;
    }

    // Handle media uploads if needed
    const _mediaIds: string[] = [];

    if (mediaData && mediaData.length > 0) {
      logger.warn("Media upload not currently supported with the modern X API");
    }

    const result = await sendPost(this.client, text, mediaData);

    // Add to recent posts cache to prevent future duplicates
    await addToRecentPosts(this.runtime, username, text);

    return result;
  }
}

File: plugins/plugin-xai/python/elizaos_plugin_xai/actions/post.py
from __future__ import annotations

import re
from dataclasses import dataclass
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from elizaos.types.components import ActionResult, HandlerCallback, HandlerOptions
    from elizaos.types.memory import Memory
    from elizaos.types.runtime import IAgentRuntime
    from elizaos.types.state import State

import logging

from elizaos.types.components import Action, ActionExample
from elizaos.types.model import ModelType

from elizaos_plugin_xai.client import TwitterClient

logger: logging.Logger | None = None


def _get_logger() -> logging.Logger:
    global logger
    if logger is None:
        from elizaos.logger import create_logger

        logger = create_logger(__name__)
    return logger


@dataclass
class PostActionResult:
    success: bool
    text: str | None = None
    error: str | None = None
    post_id: str | None = None
    post_url: str | None = None


async def validate_post(
    runtime: IAgentRuntime,
    _message: Memory | None = None,
    _state: State | None = None,
) -> bool:
    try:
        service = runtime.get_service("x")
        return service is not None
    except Exception:
        return False


async def handle_post(
    runtime: IAgentRuntime,
    message: Memory,
    _state: State | None = None,
    _options: HandlerOptions | None = None,
    callback: HandlerCallback | None = None,
    _responses: list[Memory] | None = None,
) -> ActionResult:
    log = _get_logger()
    log.info("Executing POST action")

    try:
        service = runtime.get_service("x")
        if not service:
            raise RuntimeError("X service not available")

        if not hasattr(service, "client") or not isinstance(service.client, TwitterClient):
            raise RuntimeError("X client not available or invalid")

        client: TwitterClient = service.client

        try:
            profile = await client.me()
        except Exception as e:
            raise RuntimeError(f"X client not initialized - {e}") from e

        text = ""
        if message.content:
            if isinstance(message.content, dict):
                text = message.content.get("text", "")
            elif hasattr(message.content, "text"):
                text = message.content.text or ""
            else:
                text = str(message.content)

        text = text.strip() if text else ""

        if not text:
            error_msg = "I need something to post! Please provide the text."
            if callback:
                await callback({"text": error_msg, "action": "POST"})
            return {"success": False, "error": "No text provided"}

        if len(text) > 280:
            sentences = re.findall(r"[^.!?]+[.!?]+", text) or [text]
            truncated = ""
            for sentence in sentences:
                if len(truncated + sentence) <= 280:
                    truncated += sentence
                else:
                    break
            text = truncated.strip() or f"{text[:277]}..."

        final_text = text
        if len(text) < 50 or "post" in text.lower():
            character = runtime.character
            topics = ", ".join(character.topics) if character.topics else "technology, AI"

            prompt = f"""You are {character.name}.
{character.bio or ""}

Generate a post based on: {text}

Style:
- Be specific, opinionated, authentic
- No generic content or platitudes
- Share insights, hot takes, unique perspectives
- Conversational and punchy
- Under 280 characters
- Skip hashtags unless essential

Topics: {topics}

Post:"""

            try:
                response = await runtime.use_model(
                    ModelType.TEXT_SMALL.value,
                    {
                        "prompt": prompt,
                        "maxTokens": 100,
                        "temperature": 0.9,
                    },
                )
                final_text = str(response).strip()
            except Exception as e:
                log.warning(f"Failed to generate post text: {e}, using original text")
                final_text = text

        result = await client.create_post(final_text)

        if result and result.id:
            post_id = result.id
            post_url = f"https://x.com/{profile.username}/status/{post_id}"

            log.info(f"Posted: {post_id}")

            from elizaos.types.memory import Content

            await runtime.create_memory(
                Memory(
                    entity_id=runtime.agent_id,
                    agent_id=runtime.agent_id,
                    room_id=message.room_id,
                    content=Content(
                        text=final_text,
                        url=post_url,
                        source="x",
                        action="POST",
                    ),
                ),
                "messages",
            )

            success_text = f'Posted: "{final_text}"\n\n{post_url}'
            if callback:
                await callback(
                    {
                        "text": success_text,
                        "metadata": {"postId": post_id, "postUrl": post_url},
                    }
                )

            return {
                "success": True,
                "text": f"Posted: {post_url}",
                "data": {"postId": post_id, "postUrl": post_url},
            }
        else:
            raise RuntimeError("Failed to post - no response data")

    except Exception as e:
        log.error(f"Error in POST action: {e}")
        error_msg = f"Failed to post: {e}"
        if callback:
            await callback({"text": error_msg, "action": "POST"})
        return {"success": False, "error": str(e)}


# Action definition
POST_ACTION: Action = {
    "name": "POST",
    "similes": ["POST_TO_X", "POST", "SEND_POST", "SHARE_ON_X"],
    "description": "Post content on X (formerly Twitter)",
    "validate": validate_post,
    "handler": handle_post,
    "examples": [
        [
            ActionExample(
                name="{{user1}}",
                content={"text": "Post about the weather today"},
            ),
            ActionExample(
                name="{{agent}}",
                content={"text": "I'll post about today's weather.", "action": "POST"},
            ),
        ],
        [
            ActionExample(
                name="{{user1}}",
                content={"text": "Post: The future of AI is collaborative intelligence"},
            ),
            ActionExample(
                name="{{agent}}",
                content={"text": "I'll post that for you.", "action": "POST"},
            ),
        ],
    ],
}
